content/posts/cloudbees-core-on-kubernetes-best-practices/core-psp.md

```diff
@@ -0,0 +1,234 @@
+
+Now that PSPs are applied to all the necessary `Roles` and `ClusterRoles` you can enable the Pod Security Policy Admission Controller for your GKE cluster:
+```shell
+gcloud beta container clusters update [CLUSTER_NAME] --zone [CLUSTER_ZONE] --enable-pod-security-policy
+```
+
+Next, you should ensure that all `pods` are still running:
+```shell
+kubectl get pods --all-namespaces
+```
+If a `pod` that you expect to be running is not, you need to find the `Role`/`ClusterRole` that is used for the `pod`/`deployment`/`service` and apply a PSP to it.
+
+
+Default Pod Security Policies created when enabling the `pod-security-policy` feature on a GKE cluster:
+```shell
+NAME                           PRIV    CAPS   SELINUX    RUNASUSER   FSGROUP    SUPGROUP   READONLYROOTFS   VOLUMES
+gce.event-exporter             false          RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            hostPath,secret
+gce.fluentd-gcp                false          RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            configMap,hostPath,secret
+gce.persistent-volume-binder   false          RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            nfs,secret
+gce.privileged                 true    *      RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            *
+gce.unprivileged-addon         false          RunAsAny   RunAsAny    RunAsAny   RunAsAny   false            emptyDir,configMap,secret
+```
+
+>NOTE: The default Pod Security Policies created automatically cannot be modified - Google will automatically change them back to those above.
+
+[AWS EKS](https://docs.aws.amazon.com/eks/latest/userguide/pod-security-policy.html) and [Azure AKS - Preview](https://docs.microsoft.com/en-us/azure/aks/use-pod-security-policies) also support Pod Security Policies.
+
+## Oh no, My Jenkins Agents Won't Start!
+The [Jenkins Kubernetes plugin](https://github.com/jenkinsci/kubernetes-plugin) (for ephemeral K8s agents) defaults to using a K8s [`emptyDir` volume](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) type for the Jenkins agent workspace. This causes issues when using a restrictive PSP such at the **cb-restricted** PSP above. Kubernetes defaults to mounting `emptyDir` volumes as `root:root` with permissions set to `750` - as [detailed by this GitHub issue](https://github.com/kubernetes/kubernetes/issues/2630) opened way back in 2014. When using a PSP, with Jenkins K8s agent pods, that doesn't allow containers to run as `root` the containers will not be able to access the default K8s plugin workspace directory. One approach for dealing with this is to set the K8s `securityContext` for `containers` in the `pod` spec. You can do this in the K8s plugin UI via the **Raw yaml for the Pod** field:
```

Comment:
"... that doesn't allow containers to run as `root` the containers... "

So here you can relate it if mentioned the privileged container details. I think it's useful to understand the reason of PSP enforcements.

Reply:
Added reasons to PSP yaml - inline.